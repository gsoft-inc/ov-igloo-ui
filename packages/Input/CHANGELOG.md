# @igloo-ui/input

## 1.1.0

### Minor Changes

- 4a388ae: Add Postfix icon position in Input component

## 1.0.7

### Patch Changes

- 8bc7f2c: remove focus border on disabled state

## 1.0.6

### Patch Changes

- 38a4994: Update dependency on @igloo-ui/icons and other @igloo libs

## 1.0.5

### Patch Changes

- 0a86f13: Update content of README

## 1.0.4

### Patch Changes

- 8c8686e: Change @import of font.css file for a @use in SASS files

## 1.0.3

### Patch Changes

- 4a256e3: IE11 compatibility

## 1.0.2

### Patch Changes

- 519f009: Uncontrolled input bugfix

## 1.0.1

### Patch Changes

- 2da8010: Fix render of elements with the css var --grey-{...}

## 1.0.0

### Major Changes

- ef50a1b: BRAKING CHANGE

  - Replace `readOnly` props by `disabled`

  New Features

  - Add options for passing a Icon on the input

  Update

  - Upgrade tokens version

## 0.1.0

### Minor Changes

- 8e6a4ec: ## Input component

  The Input component specifies an input field where the user can enter data.

  ### Component API

  | name      | description                                          |
  | --------- | ---------------------------------------------------- |
  | type      | Specifies the type to display                        |
  | isCompact | True for a compact appearance.                       |
  | value     | Specifies the value inside the input.                |
  | autoFocus | True if you need the input to be focus on page load. |
  | readOnly  | True if you need the input to be readonly.           |
  | onChange  | Function called when a key is pressed.               |
  | dataTest  | Add a data-test tag for automated tests              |
