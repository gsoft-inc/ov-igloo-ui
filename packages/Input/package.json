{
    "name": "@igloo-ui/input",
    "version": "2.3.2",
    "main": "dist/Input.js",
    "module": "dist/Input.js",
    "types": "dist/Input.d.ts",
    "styles": "dist/input.css",
    "publishConfig": {
        "access": "public"
    },
    "repository": "git+https://github.com/workleap/ov-igloo-ui.git",
    "author": "Officevibe Inc.",
    "contributors": [
        "Sylvester sylvester.vuong@gsoft.com"
    ],
    "license": "Apache-2.0",
    "peerDependencies": {
        "react": ">= 16.8.6"
    },
    "files": [
        "dist"
    ],
    "scripts": {
        "build": "rollup -c rollup.config.js"
    },
    "dependencies": {
        "@igloo-ui/tokens": "^2.1.0",
        "@hopper-ui/tokens": "^4.3.2",
        "classnames": "^2.3.2",
        "graphemer": "^1.4.0"
    },
    "devDependencies": {
        "@igloo-ui/icons": "^1.12.2"
    },
    "browserslist": [
        "> 1%",
        "last 2 versions",
        "IE >= 11"
    ]
}
