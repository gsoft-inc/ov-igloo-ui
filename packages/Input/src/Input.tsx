import React, { useEffect } from 'react';
import cx from 'classnames';

import './input.css';

const KeyCodes = {
  Comma: 44,
  Minus: 45,
  Nine: 57,
  Period: 46,
  Plus: 43,
  Zero: 48,
};

export type InputSize = 'small';

export type InputType = 'email' | 'text' | 'password' | 'number';

export interface InputProps extends React.ComponentPropsWithRef<'input'> {
  // Add class names to the surrounding DOM container.
  className?: string;
  // Form.ValidatedField state. True if has error.
  error?: boolean;
  // Specifies the type to display
  type?: InputType;
  // Input size.
  inputSize?: InputSize;
  // Specifies the value inside the input.
  value?: string | number;
  // True if you need the input to be focus on page load.
  autoFocus?: boolean;
  // True if you need the input to be readonly.
  readOnly?: boolean;
  // Function called when the value changes.
  onChange?: (e: any) => void;
  // Add a data-test tag for automated tests
  dataTest?: string;
}

const Input: React.FunctionComponent<InputProps> = React.forwardRef(
  (props: InputProps, ref: React.Ref<any>) => {
    const {
      className,
      error,
      type = 'text',
      inputSize,
      value,
      autoFocus,
      readOnly,
      onChange,
      dataTest,
      ...rest
    } = props;

    const classes = cx(
      'ids-input',
      className,
      inputSize && `ids-input--${inputSize}`,
      error,
      {
        'ids-input--readonly': readOnly,
      }
    );

    useEffect(() => {
      const refObject = ref as React.RefObject<HTMLInputElement>;
      if (autoFocus && refObject && refObject.current) {
        refObject.current.focus();
      }
    }, [ref, autoFocus]);

    const handleOnChange = (e: any) => {
      if (onChange) {
        onChange(e);

        return;
      }

      // Firefox doesn't support type="number", so here's a workaround
      // Input type="number" on Chrome supports "+", "-", "," et "."
      if (type === 'number') {
        const { charCode } = e.charCode;

        if (
          charCode !== KeyCodes.Comma &&
          charCode !== KeyCodes.Period &&
          charCode !== KeyCodes.Minus &&
          charCode !== KeyCodes.Plus &&
          (charCode < KeyCodes.Zero || charCode > KeyCodes.Nine)
        ) {
          e.preventDefault();
        } else if (
          (charCode === KeyCodes.Comma || charCode === KeyCodes.Period) &&
          value &&
          typeof value === 'string' &&
          (value.indexOf(',') !== -1 || value.indexOf('.') !== -1)
        ) {
          e.preventDefault();
        } else if (
          (charCode === KeyCodes.Minus || charCode === KeyCodes.Plus) &&
          value &&
          typeof value === 'string' &&
          (value.indexOf('-') !== -1 || value.indexOf('+') !== -1)
        ) {
          e.preventDefault();
        }
      }
    };

    return (
      <input
        ref={ref}
        className={classes}
        type={type}
        value={value}
        readOnly={readOnly}
        onChange={handleOnChange}
        data-test={dataTest}
        {...rest}
      />
    );
  }
);

export default Input;
