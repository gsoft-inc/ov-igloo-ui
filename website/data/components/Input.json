[
  [
    {
      "tags": {},
      "filePath": "/Users/franck.gaudin/Devel/Igloo/components/packages/Input/src/Input.tsx",
      "description": "",
      "displayName": "Input",
      "methods": [],
      "props": {
        "className": {
          "defaultValue": null,
          "description": "Add class names to the surrounding DOM container.",
          "name": "className",
          "parent": {
            "fileName": "components/packages/Input/src/Input.tsx",
            "name": "InputProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Input/src/Input.tsx",
              "name": "InputProps"
            }
          ],
          "required": false,
          "type": { "name": "string" }
        },
        "error": {
          "defaultValue": null,
          "description": "Form.ValidatedField state. True if has error.",
          "name": "error",
          "parent": {
            "fileName": "components/packages/Input/src/Input.tsx",
            "name": "InputProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Input/src/Input.tsx",
              "name": "InputProps"
            }
          ],
          "required": false,
          "type": { "name": "boolean" }
        },
        "type": {
          "defaultValue": null,
          "description": "Specifies the type to display",
          "name": "type",
          "parent": {
            "fileName": "components/packages/Input/src/Input.tsx",
            "name": "InputProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Input/src/Input.tsx",
              "name": "InputProps"
            }
          ],
          "required": false,
          "type": { "name": "InputType" }
        },
        "isCompact": {
          "defaultValue": null,
          "description": "True for a compact appearance.",
          "name": "isCompact",
          "parent": {
            "fileName": "components/packages/Input/src/Input.tsx",
            "name": "InputProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Input/src/Input.tsx",
              "name": "InputProps"
            }
          ],
          "required": false,
          "type": { "name": "boolean" }
        },
        "value": {
          "defaultValue": null,
          "description": "Specifies the value inside the input.",
          "name": "value",
          "parent": {
            "fileName": "components/packages/Input/src/Input.tsx",
            "name": "InputProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Input/src/Input.tsx",
              "name": "InputProps"
            }
          ],
          "required": false,
          "type": { "name": "string | number" }
        },
        "autoFocus": {
          "defaultValue": null,
          "description": "True if you need the input to be focus on page load.",
          "name": "autoFocus",
          "parent": {
            "fileName": "components/packages/Input/src/Input.tsx",
            "name": "InputProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Input/src/Input.tsx",
              "name": "InputProps"
            }
          ],
          "required": false,
          "type": { "name": "boolean" }
        },
        "disabled": {
          "defaultValue": null,
          "description": "True if you need the input to be readonly.",
          "name": "disabled",
          "parent": {
            "fileName": "components/packages/Input/src/Input.tsx",
            "name": "InputProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Input/src/Input.tsx",
              "name": "InputProps"
            }
          ],
          "required": false,
          "type": { "name": "boolean" }
        },
        "onChange": {
          "defaultValue": null,
          "description": "Function called when the value changes.",
          "name": "onChange",
          "parent": {
            "fileName": "components/packages/Input/src/Input.tsx",
            "name": "InputProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Input/src/Input.tsx",
              "name": "InputProps"
            }
          ],
          "required": false,
          "type": { "name": "(e: any) => void" }
        },
        "dataTest": {
          "defaultValue": null,
          "description": "Add a data-test tag for automated tests",
          "name": "dataTest",
          "parent": {
            "fileName": "components/packages/Input/src/Input.tsx",
            "name": "InputProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Input/src/Input.tsx",
              "name": "InputProps"
            }
          ],
          "required": false,
          "type": { "name": "string" }
        },
        "prefixIcon": {
          "defaultValue": null,
          "description": "Use a prefix for add an icon before the input text",
          "name": "prefixIcon",
          "parent": {
            "fileName": "components/packages/Input/src/Input.tsx",
            "name": "InputProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Input/src/Input.tsx",
              "name": "InputProps"
            }
          ],
          "required": false,
          "type": { "name": "ReactNode" }
        },
        "suffixIcon": {
          "defaultValue": null,
          "description": "Use a suffix for add an icon after the input text",
          "name": "suffixIcon",
          "parent": {
            "fileName": "components/packages/Input/src/Input.tsx",
            "name": "InputProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Input/src/Input.tsx",
              "name": "InputProps"
            }
          ],
          "required": false,
          "type": { "name": "ReactNode" }
        }
      }
    }
  ]
]
