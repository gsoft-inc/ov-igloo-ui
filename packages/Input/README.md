# Input

Input enable the user to interact with and used for freeform data entry.

<ReferenceLinks />

## Installation

To install `@igloo-ui/input` in your project, you will need to run the following command using [npm](https://www.npmjs.com/):

```bash
npm install @igloo-ui/input
```

If you prefer [Yarn](https://classic.yarnpkg.com/en/), use the following command instead:

```bash
yarn add @igloo-ui/input
```

## Usage

Then to use the component in your code just import it!

```jsx
import Input from '@igloo-ui/input';

<Input type="text" placeholder="input type text" />;
```
