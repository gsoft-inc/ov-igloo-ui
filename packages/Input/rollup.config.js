import * as pkg from './package.json';
import { createRollupConfig } from '../../rollup.config.js';

const { name } = pkg;

export default [createRollupConfig(name)];
