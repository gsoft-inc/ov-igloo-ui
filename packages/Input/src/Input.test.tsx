/**
 * @jest-environment jsdom
 */
import React from 'react';
import { shallow } from 'enzyme';

import Input from './Input';

describe('Input', () => {
  const component = shallow(<Input>Hello world</Input>);
  test('It should render without errors', () => {
    const wrapper = component.find('.ids-input');
    expect(wrapper.length).toBe(1);
  });

  test('It should render a snapshot', () => {
    expect(component).toMatchSnapshot();
  });
});
