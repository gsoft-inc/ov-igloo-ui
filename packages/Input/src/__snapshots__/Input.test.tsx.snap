// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input It should render a number type 1`] = `
<DocumentFragment>
  <input
    class="ids-input"
    data-test="ids-input"
    type="number"
    value="12345"
  />
</DocumentFragment>
`;

exports[`Input It should render a password type 1`] = `
<DocumentFragment>
  <input
    class="ids-input"
    data-test="ids-input"
    type="password"
    value="this is a great password!"
  />
</DocumentFragment>
`;

exports[`Input It should render a snapshot 1`] = `
<DocumentFragment>
  <input
    class="ids-input"
    data-test="ids-input"
    type="text"
    value=""
  />
</DocumentFragment>
`;

exports[`Input It should render a text type 1`] = `
<DocumentFragment>
  <input
    class="ids-input"
    data-test="ids-input"
    type="text"
    value="potato"
  />
</DocumentFragment>
`;

exports[`Input It should render an email type 1`] = `
<DocumentFragment>
  <input
    class="ids-input"
    data-test="ids-input"
    type="email"
    value="hashbrown@potatocatinc.com"
  />
</DocumentFragment>
`;
